### PR TITLE
refactor: convert httpclient provider to bindgen

### PR DESCRIPTION
**Pull request by vados-cosmonic**
## Feature or Problem

This PR converts the `httpclient` provider to use `provider-wit-bindgen`, utilizing a couple features and bugfixes from other PRs:

https://github.com/wasmCloud/wasmCloud/pull/963
https://github.com/wasmCloud/wasmCloud/pull/983

Those PRs *should* land before this one, but it's not absolutely required.

## Related Issues

#743 
#680 

## Testing

Built on platform(s)
- [x] x86_64-linux
- [ ] aarch64-linux
- [ ] x86_64-darwin
- [ ] aarch64-darwin
- [ ] x86_64-windows

Tested on platform(s)
- [x] x86_64-linux
- [ ] aarch64-linux
- [ ] x86_64-darwin
- [ ] aarch64-darwin
- [ ] x86_64-windows

### Acceptance or Integration

Pending implementation of `wasi:outgoing-handler` in host

### Manual Verification

Tested locally

### CODE REVIEW

**Review by rvolosatovs — CHANGES_REQUESTED** [review]
requesting changes to prevent (accidental) merge until we have https://github.com/wasmCloud/wasmCloud/issues/680 done

Regarding the backwards-incompatibility, I expected we would serialize lists of tuples as maps (that could be a parameter to the macro invocation or we can just hardcode it for now)

**Comment by rvolosatovs** [comment]
> As of 2023/10/13, this code doesn't work due to conversions between `map` (which doesn't exist in WIT) as expressed by the Smithy contract and the `list<tuple<string, list<u8>>>` that WIT has to take. The error looks like this:
> 
> ```
> {"error":"actor: sending req: Host send error failed to call `wasmcloud:bus/host.call`: provider call failed: Error when handling invocation: Error when deserializing invocation: Syntax(\"invalid type: map, expected a sequence\")"}
> ```
> 
> This should be resolved by #680, as the host will likely have to munge invocations and responses to translate maps into/out of what WIT can actually understand.

#680 is not expected to be anyhow WIT-aware, message format should be fully backwards-compatible, meaning that `wasi:http/outgoing-handler` will be sending msgpack maps, just like before. Bindgen, however, should be able to handle that and decode a msgpack map into a list of tuples and other way round

File: crates/providers/http-client/Cargo.toml
Comment (vados-cosmonic) [reply crates/providers/http-client/Cargo.toml:19]: Thanks for the catch, removed these unused crates

**Comment by vados-cosmonic** [comment]
> https://github.com/wasmCloud/wasmCloud/issues/680 is not expected to be anyhow WIT-aware

Well glad I brought this up early -- could you expand on how #680 is not WIT-aware? Maybe the scope of work isn't well defined, there's not much in the issue, but `wasi:http/outgoing-handler` is an intrinsically WIT-based concept.

I thought that the work represented by that issue was implementing the WIT interfaces that are specified by `wasi:http/outgoing-handler` so they can be used by wasmCloud actors?

> , message format should be fully backwards-compatible, meaning that wasi:http/outgoing-handler will be sending msgpack maps, just like before. Bindgen, however, should be able to handle that and decode a msgpack map into a list of tuples and other way round

While it is *possible* for bindgen to generate an object that converts to a msgpack map, it's not clear that it *should* be there, rather than in the host which is doing the `wasi:http/outgoing-handler` to `HttpClient.<operation>` conversions. 

The host intercepts every call, for every operation and does some transformations already.

The reason we *know* that the value should be a map is because of the contract. These kinds of transformations are [already in the code](https://github.com/wasmCloud/wasmCloud/blob/main/crates/host/src/wasmbus/mod.rs#L486), so could you explain why this time should be different?

Even if it was something that should be totally handled in WIT at the bindgen level, it's not safe to convert every `list<tuple<string, list<u8>>>` into a `HashMap` (or whatever type we choose to serialize to a msgpack map), as it's possible that the intended value *is* a list of tuples. 

WIT doesn't currently have a way of telling you "this thing should have been a HashMap for Smithy", that's a requirement that we're injecting by trying to make WIT speak across Smithy interfaces. 

For example, if WIT was talking to WIT, both sides would expect the list of tuples, not a map.

**Comment by vados-cosmonic** [comment]
> meaning that wasi:http/outgoing-handler will be sending msgpack maps, just like before. 

This seems to be a bit confusingly worded -- `wasi:http/outgoing-handler` *cannot* send out msgpack maps, because WIT doesn't have a map type. Is what you mean here is that #680 will make sure that the *host* intercepts `wasi:http/outgoing-handler` calls and converts them into objects that serialize as msgpack maps?

> Bindgen, however, should be able to handle that and decode a msgpack map into a list of tuples and other way round

This is going to require some deeper changes to bindgen -- it's not clear that is safe to turn every `list<tuple<string, list<u8>>>` or any such type into something that serializes to a type that serializes to a msgpack map, and this would break normal WIT <-> WIT semantics. 

This will likely have to be some kind of configuration option.

**Comment by rvolosatovs** [comment]
> > meaning that wasi:http/outgoing-handler will be sending msgpack maps, just like before.
> 
> This seems to be a bit confusingly worded -- `wasi:http/outgoing-handler` _cannot_ send out msgpack maps, because WIT doesn't have a map type. Is what you mean here is that #680 will make sure that the _host_ intercepts `wasi:http/outgoing-handler` calls and converts them into objects that serialize as msgpack maps?

yes

> > Bindgen, however, should be able to handle that and decode a msgpack map into a list of tuples and other way round
> 
> This is going to require some deeper changes to bindgen -- it's not clear that is safe to turn every `list<tuple<string, list<u8>>>` or any such type into something that serializes to a type that serializes to a msgpack map, and this would break normal WIT <-> WIT semantics.
> 
> This will likely have to be some kind of configurable option.

Yes, ideally it's a parameter to the macro, as outlined in my original comment https://github.com/wasmCloud/wasmCloud/pull/745#pullrequestreview-1674510237, but for now we can just hard-code it, because we are the only users of this crate

File: crates/providers/Cargo.toml
Comment (rvolosatovs) [inline crates/providers/Cargo.toml:35]: still looks unused to me

Comment (vados-cosmonic) [reply crates/providers/Cargo.toml:35]: Removed!

**Comment by vados-cosmonic** [comment]
Thanks for the clarifications -- I'm not too keen on hard-coding it, so I'll take a little time to think about a reasonable way to do it flexibly.

**Comment by rvolosatovs** [comment]
@vados-cosmonic any idea why the provider build artifacts might have changed in this PR?
Does e.g.
```
cargo clippy --release -j 2 --all-targets --workspace -- --deny warnings
```
work locally for you with this change set?

**Comment by vados-cosmonic** [comment]
Not sure, but it looks like the `test-providers` build in `build.rs` is failing to build after the rebase -- will look into it and see what changed.

**Comment by rvolosatovs** [comment]
looks like this is blocked on #869 , but I have no permissions to approve/merge, so we'll have to wait until someone wakes up in US

**Comment by rvolosatovs** [comment]
feel free to just commit the regenerated `Cargo.lock`

**Comment by brooksmtownsend** [comment]
@rvolosatovs / @vados-cosmonic merged that PR!

**Review by brooksmtownsend — COMMENTED** [review]
@rvolosatovs This looks good to me at a high level! Let me know if you need an additional approval here to merge but leaving it to you since you've been following

**Comment by vados-cosmonic** [comment]
Hey @brooksmtownsend this isn't ready to merge yet -- I still haven't been able to get the tests working even after rebasing -- custom build isn't passing still even with the `Cargo.lock` changes.

Regardless of that, given that the host implementation changed there will likely be changes required here.

Going to put this back in Draft so people aren't reviewing it.

File: crates/providers/http-client/src/lib.rs
Comment (connorsmith256) [inline crates/providers/http-client/src/lib.rs:43]: Nit: 
```suggestion
            ProviderInvocationError::Provider(format!("failed to convert method: {}:{}", &req.method, e))
```

File: crates/providers/http-client/src/lib.rs
Comment (connorsmith256) [inline crates/providers/http-client/src/lib.rs:58]: It looks like this was migrated from the existing code, but I'm trying to figure out what "needs to be tested" means. Surely we've checked this behavior? 😅

File: crates/providers/http-client/src/lib.rs
Comment (connorsmith256) [inline crates/providers/http-client/src/lib.rs:74]:
```suggestion
                ProviderInvocationError::Provider(format!("error receiving response body: {}", e))
```

File: crates/providers/http-client/src/lib.rs
Comment (connorsmith256) [inline crates/providers/http-client/src/lib.rs:63]:
```suggestion
                ProviderInvocationError::Provider(format!("error sending request: {}", e))
```

**Review by connorsmith256 — CHANGES_REQUESTED** [review]
One blocking request, and a few nits. I'm assuming you'll split out the bindgen code from this commit once https://github.com/wasmCloud/wasmCloud/pull/963 is merged?

File: crates/providers/http-client/src/bin/httpclient.rs
Comment (vados-cosmonic) [reply crates/providers/http-client/src/bin/httpclient.rs:4]: Sure, will do! I do think that people would want to be able to configure the OTEL stuff on their providers (possibly writing  out to a different place), but we'll cross that bridge when we get there I guess!

- [ ] remove `tracing_subscriber::fmt().init()`

File: crates/providers/http-client/src/lib.rs
Comment (vados-cosmonic) [reply crates/providers/http-client/src/lib.rs:74]: Went with a slightly different message here, but thanks for this!

File: crates/providers/http-client/src/lib.rs
Comment (vados-cosmonic) [reply crates/providers/http-client/src/lib.rs:63]: Fixed this with a slightly different message!

File: crates/providers/http-client/src/lib.rs
Comment (vados-cosmonic) [reply crates/providers/http-client/src/lib.rs:58]: Yeah I assume this is how the behavior was found... Maybe worth trying to add a test that does this explicitly in a follow up?

**Review by connorsmith256 — APPROVED** [review]

File: crates/providers/http-client/src/bin/httpclient.rs
Comment (connorsmith256) [reply crates/providers/http-client/src/bin/httpclient.rs]: Do you think people want to use different otel settings for providers vs hosts? The providers currently inherit their otel settings from the host settings

Comment (vados-cosmonic) [reply crates/providers/http-client/src/bin/httpclient.rs]: Yeah I would imagine that some providers are more important than others and you'd want to know more about one in particular. 

Like I'm thinking of specifically someone running a provider that they wrote w/ the existing host and wanting to shuttle their o11y outside to some other service to see how this provider was doing specifically. I think of it like the host is mostly boring and the providers are where more action happens -- and they would change more frequently.

That said, it's totally reasonable to hook the external o11y solution to the host and filter out for the provider too... Definitely a "we can cross that bridge when we get there" thing.

File: crates/providers/http-client/src/lib.rs
Comment (vados-cosmonic) [reply crates/providers/http-client/src/lib.rs:62]: Made [an issue](https://github.com/wasmCloud/wasmCloud/issues/995) for this

**Comment by vados-cosmonic** [comment]
Ah sorry @connorsmith256 refactoring bug made it's way in there... hopefully you don't have to re-approve (I just rebased to one commit again)

File: crates/providers/http-client/src/bin/httpclient.rs
Comment (connorsmith256) [reply crates/providers/http-client/src/bin/httpclient.rs]: IMO single-provider-only traces aren't very useful, since you won't see anything preceding or following those spans. IME distributed tracing provides the most value when you have end-to-end traces

But if we're thinking of just logs, I do see what you mean. Do you want to file an issue related to this? I'll leave this thread for you to resolve either way

**Comment by thomastaylor312** [comment]
@connorsmith256 asked me to merge this for him

**Pull request merged**
